src/replication.c

```diff
@@ -1715,58 +1714,70 @@ void readSyncBulkPayload(connection *conn) {
+                              REDISMODULE_SUBEVENT_REPL_ASYNC_LOAD_STARTED,
+                              NULL);
+    } else {
+        /* Replica starts to apply data from new master, we must discard the cached
+        * master structure. */
+        serverAssert(server.master == NULL);
+        replicationDiscardCachedMaster();
 
-    /* We want our slaves to resync with us as well, if we have any sub-slaves.
-     * The master already transferred us an entirely different data set and we
-     * have no way to incrementally feed our slaves after that. */
-    disconnectSlaves(); /* Force our slaves to resync with us as well. */
-    freeReplicationBacklog(); /* Don't allow our chained slaves to PSYNC. */
+        /* We want our slaves to resync with us as well, if we have any sub-slaves.
+        * The master already transferred us an entirely different data set and we
+        * have no way to incrementally feed our slaves after that. */
+        disconnectSlaves(); /* Force our slaves to resync with us as well. */
+        freeReplicationBacklog(); /* Don't allow our chained slaves to PSYNC. */
 
-    /* We call to emptyDb even in case of REPL_DISKLESS_LOAD_SWAPDB
-     * (Where disklessLoadMakeBackup left server.db empty) because we
-     * want to execute all the auxiliary logic of emptyDb (Namely,
-     * fire module events) */
-    emptyDb(-1,empty_db_flags,replicationEmptyDbCallback);
+        serverLog(LL_NOTICE, "MASTER <-> REPLICA sync: Flushing old data");
+        emptyDb(-1,empty_db_flags,replicationEmptyDbCallback);
+    }
 
     /* Before loading the DB into memory we need to delete the readable
-     * handler, otherwise it will get called recursively since
-     * rdbLoad() will call the event loop to process events from time to
-     * time for non blocking loading. */
+    * handler, otherwise it will get called recursively since
+    * rdbLoad() will call the event loop to process events from time to
+    * time for non blocking loading. */
     connSetReadHandler(conn, NULL);
+    
     serverLog(LL_NOTICE, "MASTER <-> REPLICA sync: Loading DB in memory");
     rdbSaveInfo rsi = RDB_SAVE_INFO_INIT;
     if (use_diskless_load) {
         rio rdb;
+        redisDb *dbarray;
+        int asyncLoading = 0;
+
+        if (server.repl_diskless_load == REPL_DISKLESS_LOAD_SWAPDB) {
+            /* Async loading means we continue serving read commands during full resync, and
+             * "swap" the new db with the old db only when loading is done.
+             * It is enabled only on SWAPDB diskless replication when master replication ID hasn't changed,
+             * because in that state the old content of the db represents a different point in time of the same
+             * data set we're currently receiving from the master. */
+            if (memcmp(server.replid, server.master_replid, CONFIG_RUN_ID_SIZE) == 0) {
+                asyncLoading = 1;
+            }
+            dbarray = diskless_load_tempDb->dbarray;
```

Comment:
_"current code will just fall back to disk-based loading."_

Maybe it's the late time of the day =D, but I can't see this. It should still do the diskless load, but returning LOADING status, right?

Reply:
```c
        else if (server.repl_diskless_load == REPL_DISKLESS_LOAD_SWAPDB && !moduleAllModulesHandleReplAsyncLoad()) {
            serverLog(LL_WARNING,
                "Skipping diskless-load because there are modules that are not aware of async replication.");
            enabled = 0;
        }
```
turning off `enabled` in `useDisklessLoad` means it'll be disk-based.

Reply:
Ok, so it's about the modules. That sounds a bit hard to solve.
Falling to disk-based load creates a requirement on disk space that was not there before, I see.

Is your suggestion to do a non-async swapdb, and fire the old events for backwards compatibility, or actually bring the whole mechanism of backups back?

Reply:
i don't want to bring back the old backup/restore code (too complicated to maintain both).

I was thinking that maybe in this mode (no traffic during loading), we can support the old module API without bringing back the old backups mechanism, and maybe the modules will be ok with it.
i.e. if the modules don't do RM_OpenKey or any other odd things during loading, they won't be able the tell the difference between the old backup/restore/discard approach, and the new temp/apply/discard approach?

@MeirShpilraien you can probably tell me i'm wrong.

Reply:
regarding the fact it falls back to disk-based, which it didn't before, i don't think that's a major issue.
i don't think swapdb is very commonly used anyway, and so are modules, so their combination is not a high concern.